### PR TITLE
adds generic scaler to kubectl

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What this PR does / why we need it**:
this PR adds generic scaler to kubectl. The implementation uses a polymorphic scale client capable of operating against scale subresources which can be found here https://github.com/kubernetes/client-go/tree/master/scale

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubectl/scale_test.go
Comment (p0lyn0mial) [inline pkg/kubectl/scale_test.go]: at the beginning I was trying to use the `fake` client but it turned out I couldn't pass the `version`, so I decided to build the fake client from scratch. 

The following code was copied from https://github.com/kubernetes/client-go/blob/master/scale/client_test.go#L56

**Comment by p0lyn0mial** [comment]
In general my plan is to pass `scaleclient.ScalesGetter` to `ScalerFor` method as a parameter. The last time I checked all the data were there https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/factory_object_mapping.go#L300

what about backward compatability ?
can we remove existing scalers ? (ReplicationControllerScaler, ReplicaSetScaler, JobScaler, StatefulSetScaler, DeploymentScaler)

File: pkg/kubectl/scale.go
Comment (p0lyn0mial) [inline pkg/kubectl/scale.go]: not sure if passing `expectedScale` makes sense.

**Comment by deads2k** [comment]
/assign DirectXMan12
/assign juanvallejo

Not for 1.9, but for 1.10 this would be great.

File: pkg/kubectl/scale_test.go
Comment (DirectXMan12) [reply pkg/kubectl/scale_test.go]: what, specifically, is the issue with the fake scale client?  I'd much rather fix that (the only reason we use this type of logic in the file you linked to is so that we can test the actual client code itself).

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]
Comment about the tests inline.  We can probably remove the old scalers in favor of this.

**Comment by deads2k** [comment]
> Comment about the tests inline. We can probably remove the old scalers in favor of this.

I don't think we should do it in this pull.

File: pkg/kubectl/scale.go
Comment (deads2k) [inline pkg/kubectl/scale.go]: `wait.PollImmediate` I think

File: pkg/kubectl/scale.go
Comment (deads2k) [inline pkg/kubectl/scale.go]: `wait.PollImmediate`

File: pkg/kubectl/scale.go
Comment (deads2k) [inline pkg/kubectl/scale.go]: Just based on names, it's weird that `actualScale` is being passed as `expectedScale`

File: pkg/kubectl/scale.go
Comment (deads2k) [inline pkg/kubectl/scale.go]: Check the logic on the other ones.  As I recall, if the expected.spec.replicas != actual.spec.replicas, we consider it done.

File: pkg/kubectl/scale.go
Comment (deads2k) [reply pkg/kubectl/scale.go]: > not sure if passing expectedScale makes sense.

Seems like you just want `desiredReplicas` right?

File: pkg/kubectl/scale_test.go
Comment (deads2k) [reply pkg/kubectl/scale_test.go]: > at the beginning I was trying to use the fake client but it turned out I couldn't pass the version, so I decided to build the fake client from scratch.

Based on the code here, I'm guessing some piece is missing in the fake discovery client?  Given the value of generic scale, I could see bypassing it, but we'd certainly want to chase an issue if that's the case.

Comment (p0lyn0mial) [reply pkg/kubectl/scale_test.go]: np, I'll take one more shot at it. Do you think I could follow up that one ?

File: pkg/kubectl/scale.go
Comment (p0lyn0mial) [reply pkg/kubectl/scale.go]: yeah, we could pass `desiredReplicas` as a number (uint) and then make the following condition in `scaleHasDesiredReplicas` method:

`return actualScale.Spec.Replicas == actualScale.Status.Replicas &&
			desiredReplicas == actualScale.Status.Replicas, nil`

does it make sense ?

File: pkg/kubectl/scale.go
Comment (p0lyn0mial) [reply pkg/kubectl/scale.go]: alright, notice that when the `retry` is set to `nil` we end up in two immediate calls to the server. Perhaps we could delay the second call i.e. set it to 1 second.

Comment (deads2k) [reply pkg/kubectl/scale.go]: > alright, notice that when the retry is set to nil we end up in two immediate calls to the server. Perhaps we could delay the second call i.e. set it to 1 second.

I think I'd go the other direction and avoid the second call, but as  starting point, two immediate calls is ok.

File: pkg/kubectl/scale_test.go
Comment (p0lyn0mial) [reply pkg/kubectl/scale_test.go]: If I recall correctly the problem was with the lack of `RESTMapper`, because I remember having issues with the version. During the test initialisation I registered some types in the store (scheme) but then the test framework couldn't find the exact match - this is because scale client doesn't accept the version.

**Comment by p0lyn0mial** [comment]
just created an issue PTAL
https://github.com/kubernetes/kubernetes/issues/56865

**Comment by BenTheElder** [comment]
/retest
run of pull-kubernetes-unit  #56077 was affected by the issue described at kubernetes/test-infra#5845 (comment)

File: pkg/kubectl/scale_test.go
Comment (juanvallejo) [inline pkg/kubectl/scale_test.go]: nit: no linebreak here

File: pkg/kubectl/scale.go
Comment (deads2k) [reply pkg/kubectl/scale.go]: This is still outstanding.

```go 
if actualScale.Spec.Replicas != desiredReplicas{
    return true, nil
}
```

Because it means the desired scale target has been reset by something else.

File: pkg/kubectl/scale_test.go
Comment (deads2k) [inline pkg/kubectl/scale_test.go]: putting a short, descriptive name in the scenario struct will help us remember what these are all doing and give a better name here.

```go
[] struct{
    name string
    .....
```

File: pkg/kubectl/scale_test.go
Comment (deads2k) [reply pkg/kubectl/scale_test.go]: > what, specifically, is the issue with the fake scale client? I'd much rather fix that (the only reason we use this type of logic in the file you linked to is so that we can test the actual client code itself).

@DirectXMan12 I think we'll take a little debt here for the net reduction this will bring to `kubectl` and the increase in power.



> If I recall correctly the problem was the lack of RESTMapper in the fake client, because I remember having issues with the version. During the test initialisation I registered some types in the store (scheme) but then the test framework couldn't find the exact match - this is because scale client doesn't accept the version.

@juanvallejo After this merges, can you see what you can do with it?  I suspect that a mapper of some kind in the FakeScaleClient from groupresource to version can get the job done.

**Comment by deads2k** [comment]
Fix up the `scaleHasDesiredReplicas` and the nit, then squash and this lgtm.

Nice job.

**Comment by p0lyn0mial** [comment]
bazel don't vex me

**Comment by p0lyn0mial** [comment]
@deads2k can I get lgtm one more time :)

**Pull request merged**
